src/blocks/map/edit.js

```diff
@@ -0,0 +1,384 @@
+			// Slide added
+			setCurrentItemIndex( slideBlocks.length - 1 );
+		} else if ( slideBlocks.length < slideCount.current ) {
+			// Slide deleted
+			if ( currentItemIndex + 1 > slideBlocks.length ) {
+				setCurrentItemIndex( slideBlocks.length - 1 );
+			}
+		}
+
+		// Update ref with new value..
+		slideCount.current = slideBlocks.length;
+	}, [ slideBlocks.length, currentItemIndex, slideCount, slideBlocks ] );
+
+	/**
+	 * Init the map.
+	 */
+	useEffect( () => {
+		// eslint-disable-next-line no-undef
+		if ( ! wmf?.apiKey ) {
+			return;
+		}
+
+		const fullScreenControl = new mapboxgl.NavigationControl();
+
+		if ( map ) {
+			map.removeControl( fullScreenControl );
+			map.remove();
+		}
+
+		// eslint-disable-next-line no-undef
+		mapboxgl.accessToken = wmf?.apiKey;
+		map = new mapboxgl.Map( {
+			container: 'map',
+			center: [ 2.8, 40 ],
+			projection: 'equalEarth',
+			scrollZoom: false,
+			style: mapStyle || 'mapbox://styles/mapbox/light-v11', // 'mapbox://styles/mattwatsonhm/clu09j0hw00tf01p7dpw5hyv7' >- custom grey colours.
+			zoom: 2,
+		} );
+
+		map.addControl( fullScreenControl );
+
+		map.on( 'load', () => {
+			// For clustering we need to add a data source.
+			map.addSource( 'markers', {
+				type: 'geojson',
+				data: {
+					type: 'FeatureCollection',
+					features:
+						slideBlocks?.map( ( slideBlock, index ) => {
+							const { id, lat, long } = slideBlock.attributes;
+							return {
+								geometry: {
+									type: 'Point',
+									coordinates: [ long, lat ],
+								},
+								type: 'Feature',
+								properties: {
+									clientId: slideBlock.clientId,
+									id,
+									index,
+								},
+							};
+						} ) || [],
+				},
+				cluster: true,
+				clusterRadius: 10,
+				clusterProperties: {
+					sum: [
+						'get',
+						'count',
+						[ 'get', 'value', [ 'properties' ] ],
+					],
+				},
+			} );
+
+			// For clustering we need to add a layer for markers.
+			// We will replace these with HTML markers.
+			map.addLayer( {
+				id: 'marker_circle',
+				type: 'circle',
+				source: 'markers',
+				filter: [ '!=', 'cluster', true ],
+				paint: {
+					'circle-color': '#000',
+					'circle-radius': 1,
+				},
+			} );
+
+			// For clustering we need to add a layer for clusters.
+			// We will replace these with HTML markers.
+			map.addLayer( {
+				id: 'marker_label',
+				type: 'symbol',
+				source: 'markers',
+				filter: [ '!=', 'cluster', true ],
+				layout: {
+					'text-field': '{point_count_abbreviated}',
+					'text-size': 10,
+				},
+				paint: {
+					'circle-color': '#000',
+					'text-color': 'white',
+				},
+			} );
+
+			// Rendering happen as the map moves, update the pins on the map
+			// based on the above data source.
+			map.on( 'render', () => {
+				if ( ! map.isSourceLoaded( 'markers' ) ) {
+					return;
+				}
+				updateMarkers();
+			} );
+		} );
+	}, [ mapStyle, slideBlocks, slideBlocks.length, updateMarkers ] );
```

Comment:
One thing I notice is that the map re-renders in the editor for every change, even just text changes in the slide blocks. This makes editing content really laggy, and I think it could be exhausting the API call limits. Can you limit the dependencies array so that the callback only runs once, or at most, once when markers are added or repositioned?

Reply:
I've sorted this in the story PR, good spot, thank you.